### PR TITLE
Shipment, partners, and marketplace controller

### PR DESCRIPTION
**Pull request by natehishon**
update

### CODE REVIEW

File: src/main/java/com/webservices/repositories/OrderDAOImpl.java [hunk 1]
```diff
@@ -37,6 +38,44 @@ public Order getOrder(Integer id) {
...
     }
 
+    @Override
+    public OrderStatus getOrderStatus(Integer id) {
+        Session session = sessionFactory.openSession();
+        OrderStatus orderStatus = null;
+        try {
+            orderStatus = session.get(Order.class, id).getOrderStatus();
+        } catch (Exception e) {
+            logger.error("Unable to get order with id " + id, e);
```
Comment (mDemianchuk) [inline src/main/java/com/webservices/repositories/OrderDAOImpl.java]: ..to get status of the order with id ....

Comment (mDemianchuk) [reply src/main/java/com/webservices/repositories/OrderDAOImpl.java]: actually nvm :)

File: src/main/java/com/webservices/repositories/PartnerDAOImpl.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    @Override
+    public Partner getPartner(Integer id) {
+        Session session = sessionFactory.openSession();
+        Partner partner = null;
+        try {
+            partner = session.get(Partner.class, id);
+        } catch (Exception e) {
+            logger.error("Unable to get shipment with id " + id, e);
```
Comment (mDemianchuk) [inline src/main/java/com/webservices/repositories/PartnerDAOImpl.java]: ...partner with id

File: src/main/java/com/webservices/repositories/PartnerDAOImpl.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        } finally {
+            session.close();
+            return partner;
+        }
+    }
+
+    @Override
+    public Boolean pushOrderToPartner(Partner partner, Order order) {
```
Comment (mDemianchuk) [inline src/main/java/com/webservices/repositories/PartnerDAOImpl.java]: assuming that we want to add this order to an existing partner we need to actually pull a proxy of an existing partner from our db and update it

File: src/main/java/com/webservices/repositories/PartnerDAOImpl.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        Session session = sessionFactory.openSession();
+        Transaction transaction = null;
+        try {
+            transaction = session.beginTransaction();
```
Comment (mDemianchuk) [inline src/main/java/com/webservices/repositories/PartnerDAOImpl.java]: here you do `Partner existingPartner = session.get(Partner.class, partner.getId());`

File: src/main/java/com/webservices/repositories/PartnerDAOImpl.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            List<Order> orderList = partner.getOrders();
```
Comment (mDemianchuk) [inline src/main/java/com/webservices/repositories/PartnerDAOImpl.java]: ... `= existingPartner.getOrders();`

File: src/main/java/com/webservices/repositories/PartnerDAOImpl.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            orderList.add(order);
+            partner.setOrders(orderList);
+            session.save(partner);
```
Comment (mDemianchuk) [inline src/main/java/com/webservices/repositories/PartnerDAOImpl.java]: and here `session.update(existingPartner);`

File: src/main/java/com/webservices/repositories/PartnerDAOImpl.java (hunk 3 above)
Comment (mDemianchuk) [reply src/main/java/com/webservices/repositories/PartnerDAOImpl.java]: so, we don't actually need to pass entire `Partner` instance, all we need is its `id` and a new `Order`

File: src/main/java/com/webservices/services/MarketplaceServiceImpl.java [hunk 7]
```diff
@@ -75,8 +76,19 @@ public Boolean acceptPayment(Order order) {
     }
 
     @Override
-    public void shipOrder(Order order) {
-        order.setOrderStatus(OrderStatus.SHIPPED);
+    public Boolean shipOrder(Order order) {
```
Comment (mDemianchuk) [inline src/main/java/com/webservices/services/MarketplaceServiceImpl.java]: how about we either pass an `Order` and an `Address` we want to ship this order to
or we just pass a `Shipment` with already set `Order` and `Address`

File: src/main/java/com/webservices/services/MarketplaceServiceImpl.java [hunk 8]
```diff
@@ -75,8 +76,19 @@ public Boolean acceptPayment(Order order) {
...
+        try {
+            order.setOrderStatus(OrderStatus.SHIPPED);
+            Shipment shipment = new Shipment();
+            shipment.setOrder(order);
+            //need address decision logic
+            shipment.setAddress(order.getCustomer().getAddresses().get(0));
+            shipmentService.addShipment(shipment);
+            return true;
+        } catch (Exception e) {
+            logger.log(Level.ERROR, "addCustomer#", e);
```
Comment (mDemianchuk) [inline src/main/java/com/webservices/services/MarketplaceServiceImpl.java]: `logger.error("appropriate error message" + an instance of the thing that caused an error, e);`

File: src/test/java/com/webservices/repositories/PartnerDAOImplTest.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                .save(Matchers.any(Parcel.class));
+
+        partnerDAO.addPartner(partner);
+
+        Mockito.verify(sessionMock, Mockito.times(1)).save(partner);
+        Mockito.verify(transactionMock, Mockito.never()).rollback();
+    }
+
+    @Test
+    public void addPartner_WhenWhenSaveThrowsException_PartnerNotAdded() {
```
Comment (mDemianchuk) [inline src/test/java/com/webservices/repositories/PartnerDAOImplTest.java]: I noticed you have multiple `_WhenWhen...` in your tests 
just autoreplace them :)

File: src/test/java/com/webservices/services/PartnerServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+                .when(partnerDAO)
+                .getPartner(invalidId);
+
+        Partner actualPartner = partnerService.getPartner(1);
+
+        Assert.assertNull(actualPartner);
+    }
+
+    @Test
+    public void pushOrderToPartner_return_true(){
```
Comment (mDemianchuk) [inline src/test/java/com/webservices/services/PartnerServiceImplTest.java]: pushOrderToPartner_WhenOrderIsPushed_ReturnsTrue

File: src/test/java/com/webservices/services/PartnerServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        Order order = getTestOrder(1);
+        Partner partner = getTestPartner(1);
+        Assert.assertEquals(partnerService.pushOrderToPartner(order, partner), true);
```
Comment (mDemianchuk) [inline src/test/java/com/webservices/services/PartnerServiceImplTest.java]: use `Assert.true();` instead

File: src/test/java/com/webservices/services/PartnerServiceImplTest.java (hunk 10 above)
Comment (mDemianchuk) [reply src/test/java/com/webservices/services/PartnerServiceImplTest.java]: your test methods should follow this structure (all separated by one blank line):
```
// Arrange 
Order order = getTestOrder(1);
Partner partner = getTestPartner(1);

// Act
Boolean updated = partnerService.pushOrderToPartner(order, partner);

// Assert
Assert.assertTrue(updated);
```

File: src/main/java/com/webservices/controllers/OrderController.java [hunk 12]
```diff
@@ -35,6 +36,36 @@ public OrderController(OrderService orderService) {
         }
     }
 
+    @RequestMapping(value = "/get-status/{id}", method = RequestMethod.GET)
```
Comment (RobusK) [inline src/main/java/com/webservices/controllers/OrderController.java]: I see a verb here. No bueno

File: src/main/java/com/webservices/controllers/OrderController.java [hunk 13]
```diff
@@ -35,6 +36,36 @@ public OrderController(OrderService orderService) {
...
+                return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+            }
+            return new ResponseEntity<>(orderStatus, HttpStatus.OK);
+        } catch (Exception e) {
+            logger.info("Unable to get order status with id " + id, e);
+            return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
+        }
+    }
+
+    @RequestMapping(value = "/cancel-order/{id}", method = RequestMethod.GET)
```
Comment (RobusK) [inline src/main/java/com/webservices/controllers/OrderController.java]: Verb here. No Bueno. Should really be a PUT on the order with {status: CANCELLED}

**Pull request merged**